R-packages/modeltools/R/matrix.R

```diff
@@ -0,0 +1,72 @@
+#' Create a training and prediction matrices for a given ahead.
+#'
+#' @param lagged_df Data frame of lagged data.  It should have the following columns:
+#'     \itemize{
+#'     \item{`geo_value`}{Strings of geographic locations.}
+#'     \item{`time_value`}{Dates of training data.}
+#'     \item{Covariate columns}{Columns with names of the form `value-{days}:{signal}` or
+#'         `value+0:{signal} whose values correspond to `{signal}` `{days}` before `time_value`}
+#'     \item{Response columns}{Columns with names of the form `response+{n}:{response}` whose values
+#'         correspond to `{response}` `{n}` incidence period units after `time_value`.}
+#'     }
+#'     A data frame in this format can be made using `covidcast::aggregate_signals()`.
+#' @param ahead Number of incidence period units (i.e., epiweeks, days, etc.) ahead to forecast
```

Comment:
I don't think this function works correctly for any incidence period other than days. (The way `quantgen_forecaster` does it only works correctly for incidence_period = "day".)

Part of me feels that it would be much easier for this function (and other related ones) to force the user to do all these operations in days. It might be easier for the user to convert the epiweek aheads into day aheads (perhaps using a separate function), then use the day aheads in these functions. I suspect this is already what @dajmcdon is doing with anteater (forecasting 1-4 epiweek ahead on a Monday corresponds to 5, 12, 19, 26 days ahead).

Reply:
Re epiweek/days: I think this has gotten bungled across different packages. There are various iterations of `sum_to_epiweek` that don't feel quite right to me. If I'm doing everything from scratch, I would do two things:

1. Rather than summing up to epiweek, I would "average up". Best to use `modeltools::Mean(x)*7`. In my opinion, there's nothing special about the day an epiweek begins on from the standpoint of the forecasting. Any weekly seasonality is removed with the "average up" operation. This allows us to then use daily data always. There's of course lots of correlation between the observations, which reduces the effective sample size, but we were often reducing our sample size from 7 to 1 per week explicitly. `Mean(x) = mean(x, na.rm=T)` so as long as a week has one observation, we have a value. This automatically imputes missing data with the trailing mean (there are some potential issues here which should get handled by the corrections).
2. Add a simple function that determines the correct target day based on incidence period and forecast date. I did this manually as @kjytay describes. I think this is the only thing necessary to enable the "epiweek" task:
```r
get_target_ahead <- function(forecast_date, incidence_period, ahead) {
  forecast_date <- lubridate::ymd(forecast_date)
  ed <- evalcast::get_target_period(forecast_date, incidence_period, ahead)$end
  return(as.numeric(ed - forecast_date))
}
```

Reply:
It doesn't matter what the incidence period is for this function because the meaning of the response here is not important.  We only need to match `ahead` in the `"response+{ahead}:{signal}"` column name to the `ahead` value passed in as an argument to this function.  Obviously Garbage In = Garbage Out, but it is not the job of this function to verify that the `reponse*` columns were created in a sensible way.

Reply:
Yes, so personally I would prefer that this function talk about everything in terms of days and avoid mention of "incidence period".

Reply:
All of the computation between time values is done in days here.  The only place that the `incidence_period` only matters in the way that we interpret the `reponse*` columns.  I am against translating things into days since we are currently very strongly using the `ahead = 1:4` epiweek convention all over, such as:

- How we call `get_predictions()`
- How we submit our weekly reports
- How all the researchers (especially senior ones!) talk about it

I've removed some references to `incidence_period` where I can in the documentation.

Reply:
The processing is much easier in days. Maybe (internally):
1. Convert inputs of epiweeks ahead into days ahead.
2. Grab lags.
3. If epiweeks, rewrite the names `response+{ahead in days}` back to `response+{ahead in epiweeks}`.

Reply:
I don't have too strong of an objection to this but it wouldn't change any code in this PR at all.  In fact, I don't see how the above changes would make a difference anywhere in the code.  Right now we have the following workflow (forgive the loose pseudocode):
```
add_response_columns <- function(df, ahead, incidence_period) {
  period <- get_target_period(incidence_period, ahead)
  sum_columns(df, period$start, period$end)
}

forecaster <- function(df_list) {
  df <- aggregate_signals(df_list, lags)
  for (a in ahead) {
     response_df <- add_response_columns(df, a, incidence_period)
     mat <- create_train_and_predict_matrices(response_df)
     predict(fit(mat))
}
```
If we convert everything to days "internally" I think we just end up with this pseudocode:
```
add_response_columns <- function(df, period_start, period_end) {
  sum_columns(df, period_start, period_end)
}

forecaster <- function(df_list) {
  df <- aggregate_signals(df_list, lags)
  for (a in ahead) {
     a <- a * incidence_period
     period <- get_response_period(a, incidence_period)
     response_df <- add_response_columns(df, period$start, period$end)
     mat <- create_train_and_predict_matrices(response_df)
     mat <- rename_response_columns_to_match_epiweek(mat)
     predict(fit(mat))
}
```
What is gained here/

Reply:
I agree. Seemingly nothing.

Reply:
Yes. For me it’s about making sure the documentation is clear.